stdlib/asm/math/poly512.masm

```diff
@@ -0,0 +1,6682 @@
+    movup.4
+    add
+
+    swap
+    movup.4
+    add
+    swap
+
+    movup.2
+    movup.4
+    add
+    movdn.2
+
+    movup.3
+    movup.4
+    add
+    movdn.3
+end
+
+# Given two operands ( i.e. field elements a, b ) on stack top, this routine computes c = a % b
+#
+# Expected stack state
+#
+# [b, a, ...]
+#
+# Output stack state looks like
+#
+# [c, ...] | c = a % b
+proc.mod_div
+    swap
+    u32split
+    movup.2
+    u32split
+    exec.u64::unchecked_mod
+    drop # higher 32 -bit must always be 0, because b = 12289
+end
+
+# Given four elements on stack top, this routine reduces them by applying
+# modular division by 12289 ( = Falcon Signature Algorithm's Prime Number )
+#
+# Input stack state :
+#
+# [a0, a1, a3, a3, ...]
+#
+# Operated such that
+#
+# b`i` = a`i` % 12289 | i ∈ [0..4)
+#
+# Output stack state :
+#
+# [b0, b1, b2, b3, ...]
+proc.mod_div_word
+    push.12289
+    exec.mod_div
+
+    swap
+    push.12289
+    exec.mod_div
+    swap
+
+    movup.2
+    push.12289
+    exec.mod_div
+    movdn.2
+
+    movup.3
+    push.12289
+    exec.mod_div
+    movdn.3
+end
+
+# Given an operand on stack, this routine negates the element, using modular arithmetic
+# over Falcon Digital Signature Algorithm's prime field = 12289.
+#
+# All this routine does is
+#
+# b = (0 - a) % Q
+#   = Q - a % Q | Q = 12289
+#
+# Input stack state
+#
+# [a,  ...]
+#
+# Output stack state looks like
+#
+# [b, ...] | b ∈ [0..12289)
+proc.neg
+    dup
+    push.12289
+    gt
+
+    if.true
+        push.12289
+        exec.mod_div
+    end
+
+    push.12289
+    swap
+    sub
+end
+
+# Given four elements on stack, this routine negates those, using modular arithmetic
+# over Falcon Digital Signature Algorithm's prime field = 12289.
+#
+# All this routine does is
+#
+# b`i` = (0 - a`i`) % Q
+#   = Q - a`i` % Q | Q = 12289 & i ∈ [0..4)
+#
+# Input stack state
+#
+# [a0, a1, a2, a3, ...]
+#
+# Output stack state looks like
+#
+# [b0, b1, b2, b3 ...] | b`i` ∈ [0..12289)
+proc.neg_word
+    exec.neg
+
+    swap
+    exec.neg
+    swap
+
+    movup.2
+    exec.neg
+    movdn.2
+
+    movup.3
+    exec.neg
+    movdn.3
+end
+
+# Given a field element, this routine does centered reduction using Miden VM
+# prime ( say Q ) and then reduces it using Falcon Post Quantum Digital 
+# Signature Algorithm prime ( say Q' )
+#
+# Q = 2 ^ 64 - 2 ^ 32 + 1
+# Q' = 12289
+#
+# Expected stack state
+#
+# [a, ...]
+#
+# All this routine does is
+#
+# if a > (Q >> 1):
+#   b = (a - Q) % Q'
+# else:
+#   b = a % Q'
+#
+# Final stack state looks like
+#
+# [b, ...]
+proc.reduce
+    dup
+    push.9223372034707292160
+    gt
+
+    if.true
+        push.12289
+        exec.mod_div
+
+        dup
+        push.7002
+        gte
+
+        if.true
+            sub.7002
+        else
+            push.7002
+            swap
+            sub
+
+            push.12289
+            swap
+            sub
+        end
+    else
+        push.12289
+        exec.mod_div
+    end
+end
```

Comment:
Here, instead of using nested if-else statements, it would probably be more efficient to compute both branches, and then drop the wrong results using `cdrop`.

Reply:
I tried rewriting them using `cdrop`, but that increases cost of execution, which is why I'm not sending that commit.

Reply:
How does it make this efficient ? Is it not in terms of compute cost (VM cycles) ?

Reply:
Sorry I missed this question earlier: using `if` requires creating another node in program MAST. This has an overhead of several cycles and an extra hash computation. The latter does not affect cycle count.

I think we can keep it as is for now and then see if we can optimize later.